### PR TITLE
refac: improve #424

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR improves #424 in several aspects.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/base/BUILD.bazel
Comment (ashjeong) [inline tachyon/math/base/BUILD.bazel:1]: Please also remove `invalid_operation.h` in the `base` folder!

File: tachyon/math/polynomials/univariate/univariate_polynomial_ops.h [hunk 1]
```diff
@@ -383,31 +383,40 @@ class UnivariatePolynomialOp<UnivariateDenseCoefficients<F, MaxDegree>> {
...
+      const UnivariatePolynomial<DOrS>& other) {
     DivResult<UnivariatePolynomial<D>> result;
-    CHECK(Divide(self, other, result));
-    result.remainder.coefficients_.RemoveHighDegreeZeros();
-    return result.remainder;
+    if (LIKELY(Divide(self, other, result))) {
+      result.remainder.coefficients_.RemoveHighDegreeZeros();
+      return result.remainder;
+    }
+    LOG_IF_NOT_GPU(ERROR) << "Division by zero attempted";
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/univariate_polynomial_ops.h:394]: To provide better error logging in these cases, what about changing the error log to "Division by zero attempted by mod"? (+ for all other cases in this commit if you agree)

**Review by ashjeong — CHANGES_REQUESTED** [review]

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole For [9034bd3](https://github.com/kroma-network/tachyon/pull/431/commits/9034bd3959ba0f4d4facf4365b59469bc820c7c2), there is a missing change in tachyon/cc/math/finite_fields/prime_field_unittest.cc!

File: tachyon/math/base/groups.h [hunk 2]
```diff
@@ -151,8 +148,8 @@ class MultiplicativeGroup : public MultiplicativeSemigroup<G> {
     // Invert |product|.
     // (a₁ * a₂ * ... *  aₙ)⁻¹
     std::optional<G> product_inv_opt = product.Inverse();
-    if (UNLIKELY(
-            InvalidOperation(!product_inv_opt, "Inverse of zero attempted"))) {
+    if (UNLIKELY(!product_inv_opt)) {
```
Comment (TomTaehoonKim) [inline tachyon/math/base/groups.h:151]: This case is exceptional for readability right?

File: tachyon/math/finite_fields/cubic_extension_field.h [hunk 3]
```diff
@@ -271,21 +270,20 @@ class CubicExtensionField : public CyclotomicMultiplicativeSubgroup<Derived> {
...
-    if (UNLIKELY(InvalidOperation(
-            !DoInverse(*static_cast<const Derived*>(this), ret),
-            "Inverse of zero attempted"))) {
-      return std::nullopt;
+    if (LIKELY(DoInverse(*static_cast<const Derived*>(this), ret))) {
+      LOG_IF_NOT_GPU(ERROR) << "Inverse of zero attempted";
+      return ret;
     }
-    return ret;
+    return std::nullopt;
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/cubic_extension_field.h:277]:
```suggestion
    if (LIKELY(DoInverse(*static_cast<const Derived*>(this), ret))) return ret;
    LOG_IF_NOT_GPU(ERROR) << "Inverse of zero attempted";
    return std::nullopt;
```

File: tachyon/math/finite_fields/cubic_extension_field.h [hunk 4]
```diff
@@ -271,21 +270,20 @@ class CubicExtensionField : public CyclotomicMultiplicativeSubgroup<Derived> {
...
-                                             *static_cast<Derived*>(this)),
-                                  "Inverse of zero attempted"))) {
-      return std::nullopt;
+    if (LIKELY(DoInverse(*static_cast<const Derived*>(this),
+                         *static_cast<Derived*>(this)))) {
+      LOG_IF_NOT_GPU(ERROR) << "Inverse of zero attempted";
+      return static_cast<Derived*>(this);
     }
-    return static_cast<Derived*>(this);
+    return std::nullopt;
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/cubic_extension_field.h:286]:
```suggestion
    if (LIKELY(DoInverse(*static_cast<const Derived*>(this), ret))) return ret;
    LOG_IF_NOT_GPU(ERROR) << "Inverse of zero attempted";
    return std::nullopt;
```

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_x86.h.tpl [hunk 5]
```diff
@@ -218,19 +217,21 @@ class PrimeField<_Config, std::enable_if_t<_Config::%{asm_flag}>> final
   // MultiplicativeGroup methods
   constexpr std::optional<PrimeField> Inverse() const {
     PrimeField ret;
-    if (UNLIKELY(InvalidOperation(!value_.template MontgomeryInverse<Config::kModulusHasSpareBit>(
-            Config::kModulus, Config::kMontgomeryR2, ret.value_), "Inverse of zero attempted"))) {
-      return std::nullopt;
+    if (LIKELY(value_.template MontgomeryInverse<Config::kModulusHasSpareBit>(
+            Config::kModulus, Config::kMontgomeryR2, ret.value_))) {
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/generator/prime_field_generator/prime_field_x86.h.tpl:221]: Is there a rule for putting/not-putting parentheses for single if-else code block? Maybe if it fits in a line then no parentheses and if not then put parentheses? Cuz above, I mostly saw without parentheses but ones that fit in a line.

File: tachyon/math/finite_fields/prime_field_gpu_debug.h [hunk 6]
```diff
@@ -394,7 +389,8 @@ class PrimeFieldGpuDebug final
 
   [[nodiscard]] constexpr static bool DoInverse(const PrimeFieldGpuDebug& a,
                                                 PrimeFieldGpuDebug& b) {
-    if (UNLIKELY(InvalidOperation(!a.IsZero(), "Inverse of zero attempted"))) {
+    if (UNLIKELY(!a.IsZero())) {
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/prime_field_gpu_debug.h:392]:
```suggestion
    if (UNLIKELY(a.IsZero())) {
```
Ooops can you make a separate commit for this bug fix?

**Comment by chokobole** [comment]
> @chokobole For [9034bd3](https://github.com/kroma-network/tachyon/pull/431/commits/9034bd3959ba0f4d4facf4365b59469bc820c7c2), there is a missing change in tachyon/cc/math/finite_fields/prime_field_unittest.cc!

In fact, they are required!

File: tachyon/math/polynomials/univariate/univariate_evaluation_domain.h [hunk 7]
```diff
@@ -336,10 +335,10 @@ class UnivariateEvaluationDomain : public EvaluationDomain<F, MaxDegree> {
       const std::optional<F> div =
           EvaluateVanishingPolynomial(tau) /
           (size_as_field_element_ * v_subdomain_of_tau);
-      if (UNLIKELY(InvalidOperation(!div, "Division by zero attempted"))) {
-        return std::nullopt;
+      if (LIKELY(div)) {
+        return subdomain.size_as_field_element_ * *div;
       }
-      return subdomain.size_as_field_element_ * *div;
+      return std::nullopt;
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/univariate/univariate_evaluation_domain.h:342]: Please add code here for error logging as well!

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Comment by TomTaehoonKim** [comment]
Please fix typo in [f23741b](https://github.com/kroma-network/tachyon/pull/431/commits/f23741b3143d93e05c425fad2611e9470a0e2d71). DivMode -> DivMod.

File: tachyon/math/polynomials/univariate/univariate_evaluation_domain.h (hunk 7 above)
Comment (chokobole) [reply tachyon/math/polynomials/univariate/univariate_evaluation_domain.h:342]: When I discussed this type of logging with @ashjeong  in #424, we agreed that logging errors can be useful. However, it becomes redundant if we log every instance where the division returns nullopt. This particular location seems redundant for logging.

File: tachyon/math/base/groups.h (hunk 2 above)
Comment (chokobole) [reply tachyon/math/base/groups.h:151]: Yes I think so. This is a typical case that handles error.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM
